core/src/main/scala/io/projectglow/gff/GffDataSource.scala

```diff
@@ -0,0 +1,245 @@
+
+    val dfWithMappedAttributes = convertAttributesToMap(
+      filterCommentsAndFasta(
+        csvDf
+      )
+    )
+
+    requiredSchema.foldLeft(dfWithMappedAttributes)((df, f) =>
+      f match {
+        case f if structFieldsEqualExceptNullability(f, startField) =>
+          df.withColumn(
+            f.name,
+            col(f.name).cast(f.dataType) - 1
+          )
+        case f if structFieldsEqualExceptNullability(f, endField) =>
+          df.withColumn(
+            f.name,
+            col(f.name).cast(f.dataType)
+          )
+        case f if structFieldsEqualExceptNullability(f, scoreField) |
+          structFieldsEqualExceptNullability(f, strandField) |
+          structFieldsEqualExceptNullability(f, phaseField) =>
+          df.withColumn(
+            f.name,
+            when(
+              col(f.name) =!= lit(NULL_IDENTIFIER),
+              col(f.name).cast(f.dataType)
+            ).otherwise(lit(null))
+          )
+        case f => // TODO: To be completed (also rectify column names)
+          df.withColumn(
+            f.name,
+            element_at(col(attributesMapColumnName), f.name)
+          )
+      }
+    )
+      .drop(attributesField.name, attributesMapColumnName) // TODO: Complete the option of keeping attributes field
+      .rdd
+  }
+}
+
+object GffDataSource {
+
+  def inferSchema(sqlContext: SQLContext, path: String): StructType = {
+    val spark = sqlContext.sparkSession
+    val csvDf = spark
+      .read
+      .option("sep", COLUMN_DELIMITER)
+      .schema(GffDataSource.gffCsvSchema)
+      .csv(path)
+
+
+    val attributeTags = convertAttributesToMap(
+      filterCommentsAndFasta(
+        csvDf
+      )
+    )
+      .select(attributesMapColumnName)
+      .withColumn(
+        attributesMapColumnName,
+        explode(
+          map_keys(
+            col(attributesMapColumnName)
+          )
+        )
+      )
+      .agg(
+        collect_set(
+          attributesMapColumnName
+        )
+      )
+      .collect()(0).getAs[Seq[String]](0)
+
+    // generate the schema. The field names will be case and underscore insensitive
+    val officialAttributeFields =
+      gffOfficialAttributeFields.foldLeft(Seq[StructField]()) { (s, f) =>
+        if (attributeTags
+          .map(t => deUnderscore(t.toLowerCase))
+          .contains(f.name.toLowerCase)) {
+          s :+ f
+        } else {
+          s
+        }
+      }
+
+    val remainingTags = attributeTags
+      .filter(
+        t =>
+          !officialAttributeFields
+            .map(_.name.toLowerCase)
+            .contains(deUnderscore(t.toLowerCase)) && !t.isEmpty
+      )
+
+    val unofficialAttributeFields = remainingTags.foldLeft(Seq[StructField]()) { (s, t) =>
+      s :+ StructField(t, StringType)
+    }
+
+    StructType(
+      gffBaseSchema.fields ++ officialAttributeFields ++ unofficialAttributeFields
+    )
+
+  }
+
+  def filterCommentsAndFasta(df: DataFrame): DataFrame = {
+    df.where(
```

Comment:
There's an option in the CSV reader to drop rows that don't have the right number of fields. Might save you some code here.

Reply:
As well as an option to filter out comments.

Reply:
Used filtering the comments. 
Used DROPMALFORMED. This has issues in spark 2.4.4 and before which I dealt with and documented in the code.